#CODING DOJO
Fecha creación:15/1/2022

Fecha de actualización: 15/1/2022

Esteban Simbaña

Practica Cargado y filtrado de datos

In [2]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [95]:
import pandas as pd
nombrearchivo = '/content/drive/MyDrive/Coding Dojo/Códigos/Solutions/data_sales/sales_predictions.csv'
sales_predictions_df = pd.read_csv(nombrearchivo)
sales_predictions_df1 = sales_predictions_df
sales_predictions_df

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
0              FDA15        9.300  ...  Supermarket Type1          3735.1380
1              DRC01        5.920  ...  Supermarket Type2           443.4228
2              FDN15       17.500  ...  Supermarket Type1          2097.2700
3              FDX07       19.200  ...      Grocery Store           732.3800
4              NCD19        8.930  ...  Supermarket Type1           994.7052
...              ...          ...  ...                ...                ...
8518           FDF22        6.865  ...  Supermarket Type1          2778.3834
8519           FDS36        8.380  ...  Supermarket Type1           549.2850
8520           NCJ29       10.600  ...  Supermarket Type1          1193.1136
8521           FDN46        7.210  ...  Supermarket Type2          1845.5976
8522           DRG01       14.800  ...  Supermarket Type1           765.6700

[8523 rows x 12 columns]

Como inicio, se muestran los tipos de datos para su análisis 

In [96]:
sales_predictions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB




*   Como se puede observar los datos en total recogidos serían de "8523" de los cuales en la columna de ('Item_Weight' , 'Outlet_Size') se encuentran datos faltantes 

*   Por otro lado, también podemos asegurar que cada tipo de variables corresponde lógicamente con los datos presentados en la sales_datadf



#  Datos faltantes 
Ahora se analiza los datos faltantes

In [97]:
sales_predictions_df.isnull().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

Ahora se analiza que porcentaje de datos faltantes hay respectivamente 

Porcentaje datos faltantes de 'Item_Weight'

In [99]:
item_weight_missing = pd.isnull(sales_predictions_df['Item_Weight']).sum()
item_weight_missing #se guarda en una variable el total de datos faltantes 

1463

Se dividen para el total de datos y se multiplican por "100"

In [78]:
count_iw = sales_predictions_df['Item_Weight'].count()
count_iw
#como la funcion cuenta solo los datos que tienen valor
#sumamos los datos para tener el total de datos

7060

In [100]:
total_item_weight = item_weight_missing +  count_iw
total_item_weight

8523

In [101]:
#ahora si realizamos la conversion a porcentaje
prcent_iw_missing = (item_weight_missing/total_item_weight)*100
prcent_iw_missing #17.16 es el porcentaje de valores faltantes en la columna 'Item_Weight'

17.165317376510618

Porcentaje datos faltantes en 'Outlet_Size'

In [102]:
#ahora utilizaremos un método más automatizado para mostrar el porcentaje de datos de 'Outlet_Size'
sales_predictions_df['Outlet_Size'].isna().sum()/len(sales_predictions_df)*100
#nos muestra en pantalla el porcentaje total

28.27642848762173

In [103]:
#Mostramos todos los datos faltantes en porcentaje
sales_predictions_df.isna().sum()/len(sales_predictions_df)*100

Item_Identifier               0.000000
Item_Weight                  17.165317
Item_Fat_Content              0.000000
Item_Visibility               0.000000
Item_Type                     0.000000
Item_MRP                      0.000000
Outlet_Identifier             0.000000
Outlet_Establishment_Year     0.000000
Outlet_Size                  28.276428
Outlet_Location_Type          0.000000
Outlet_Type                   0.000000
Item_Outlet_Sales             0.000000
dtype: float64


*   Observamos que la variable 'Outlet_Size' tiene un 28% de datos faltantes por lo que se debería considerar eliminar la variable si no es prioritaria 
*  Mientras tanto, la variable 'Item_Weight' se podría inferir un dato para imputarlo a los datos faltantes.




# Datos estadísticos 

In [104]:
#mostramos los datos estadísticos de la df
sales_predictions_df.describe()

Item_Weight  ...  Item_Outlet_Sales
count  7060.000000  ...        8523.000000
mean     12.857645  ...        2181.288914
std       4.643456  ...        1706.499616
min       4.555000  ...          33.290000
25%       8.773750  ...         834.247400
50%      12.600000  ...        1794.331000
75%      16.850000  ...        3101.296400
max      21.350000  ...       13086.964800

[8 rows x 5 columns]



1.   Se observa que la variable 'Item Weight' (la variable que tiene datos faltantes que podrían ser modificados) no esta tan dispersa con una variación estándar "std" de 4,64 
2.   La variable 'Oulet_Size' no se muestra por ser una variable categórica 
3. Se nota que hay una gran variación entre los datos de 'Item_Oulet_Sales' pero no hay gran problema porque tiene sentido su variación ya que pueden haber productos que se vendieron más o menos, debido a su popularidad, necesidad del cliente,etc.


# Rellenar datos faltantes 

Como se mostró anteriormente se puede modificar los datos faltantes de la variable 'Item_Weight'
Ya que es una variable de tipo númerica se puede optar por reemplazar los datos por el promedio "mean" de la variable.

In [105]:
#mostramos en una variable el promedio de la columna 'Item_Weight'
mean_Item_Weight = sales_predictions_df1['Item_Weight'].mean()
mean_Item_Weight

12.857645184136183

In [118]:
#ahora reemplazamos en sales_prediction_df1
sales_predictions_df1['Item_Weight'] = sales_predictions_df1['Item_Weight'].fillna(mean_Item_Weight)
sales_predictions_df1.head()

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
0           FDA15         9.30  ...  Supermarket Type1          3735.1380
1           DRC01         5.92  ...  Supermarket Type2           443.4228
2           FDN15        17.50  ...  Supermarket Type1          2097.2700
3           FDX07        19.20  ...      Grocery Store           732.3800
4           NCD19         8.93  ...  Supermarket Type1           994.7052

[5 rows x 12 columns]

In [119]:
#verificamos los datos faltantes
sales_predictions_df1.isna().sum()

Item_Identifier                 0
Item_Weight                     0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

# Eliminar datos faltantes


Eliminamos las filas con variables faltantes en la columna de 'Oulet_Size'

In [126]:
sales_predictions_df1 = sales_predictions_df1.dropna(subset=['Outlet_Size'])
sales_predictions_df1.head()

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
0           FDA15        9.300  ...  Supermarket Type1          3735.1380
1           DRC01        5.920  ...  Supermarket Type2           443.4228
2           FDN15       17.500  ...  Supermarket Type1          2097.2700
4           NCD19        8.930  ...  Supermarket Type1           994.7052
5           FDP36       10.395  ...  Supermarket Type2           556.6088

[5 rows x 12 columns]

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
0              FDA15        9.300  ...  Supermarket Type1          3735.1380
1              DRC01        5.920  ...  Supermarket Type2           443.4228
2              FDN15       17.500  ...  Supermarket Type1          2097.2700
4              NCD19        8.930  ...  Supermarket Type1           994.7052
5              FDP36       10.395  ...  Supermarket Type2           556.6088
...              ...          ...  ...                ...                ...
8517           FDF53       20.750  ...  Supermarket Type1          3608.6360
8518           FDF22        6.865  ...  Supermarket Type1          2778.3834
8520           NCJ29       10.600  ...  Supermarket Type1          1193.1136
8521           FDN46        7.210  ...  Supermarket Type2          1845.5976
8522           DRG01       14.800  ...  Supermarket Type1           765.6700

[6113 rows x 12 columns]

# Base de datos final 


Aplicados los cambios se mostraría la df final, teniendo en cuenta que con la poca información del negocio no podríamos borrar ninguna columna del todo ya que no conocemos a profundidad la base de datos

In [127]:
#base de datos final (aplicado un data_cleaning)

In [128]:
sales_predictions_df1

Item_Identifier  Item_Weight  ...        Outlet_Type  Item_Outlet_Sales
0              FDA15        9.300  ...  Supermarket Type1          3735.1380
1              DRC01        5.920  ...  Supermarket Type2           443.4228
2              FDN15       17.500  ...  Supermarket Type1          2097.2700
4              NCD19        8.930  ...  Supermarket Type1           994.7052
5              FDP36       10.395  ...  Supermarket Type2           556.6088
...              ...          ...  ...                ...                ...
8517           FDF53       20.750  ...  Supermarket Type1          3608.6360
8518           FDF22        6.865  ...  Supermarket Type1          2778.3834
8520           NCJ29       10.600  ...  Supermarket Type1          1193.1136
8521           FDN46        7.210  ...  Supermarket Type2          1845.5976
8522           DRG01       14.800  ...  Supermarket Type1           765.6700

[6113 rows x 12 columns]